src/NuGet.Core/NuGet.Packaging/Licenses/LicenseExpressionTokenizer.cs

```diff
@@ -0,0 +1,136 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+
+namespace NuGet.Packaging.Licenses
+{
+    internal class LicenseExpressionTokenizer
+    {
+        private readonly string _value;
+        /// <summary>
+        /// A tokenizer for a license expression.
+        /// This implementation assumes that the input has been sanitized and that there are no invalid characters.
+        /// </summary>
+        /// <param name="value"></param>
+        internal LicenseExpressionTokenizer(string value)
+        {
+            if (string.IsNullOrWhiteSpace(value))
+            {
+                throw new ArgumentException(nameof(value));
+            }
+            _value = value.Trim();
+        }
+
+        /// <summary>
+        /// The valid characters for a license identifier are a-zA-Z0-9.-+
+        /// The valid characters for a license expression are the above whitespace and ().
+        /// </summary>
+        /// <param name="value"></param>
+        /// <returns></returns>
+        internal bool HasValidCharacters()
+        {
+            for (var i = 0; i < _value.Length; i++)
+            {
+                // If the character is not among these characters
+                if (!((_value[i] >= 'a' && _value[i] <= 'z') ||
+                    (_value[i] >= 'A' && _value[i] <= 'Z') ||
+                    (_value[i] >= '0' && _value[i] <= '9') ||
+                    _value[i] == ' ' ||
+                    _value[i] == '.' ||
+                    _value[i] == '-' ||
+                    _value[i] == '+' ||
+                    _value[i] == '(' ||
+                    _value[i] == ')'
+                    ))
+                {
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        /// <summary>
+        /// Given a string, tokenizes by space into operators and values. The considered operators are, AND, OR, WITH, (, and ). 
+        /// </summary>
+        /// <returns>Tokens</returns>
+        internal IEnumerable<LicenseExpressionToken> Tokenize()
```

Comment:
Why are you running 1*(ALPHA AND (MIT OR ASDD+312 WITH gdfgg))? 

The 1* is invalid. 

Furthermore this is simply the tokenizer. 

It will only create tokens..given a valid expression namely something like 
> (0BSD AND (MIT OR Apache-2.0 WITH 389-exception))
it will generate the list of tokens. 
(, 0BSD, AND, (, MIT, OR, Apache-2.0, WITH, 389-exception, ), )

Furthermore in your example:

The * is illegal. 
The values after the + are illegal. The + operator is only valid at the end of an identifier. 

Regarding the parsing logic, it's similar to what you are articulating and the logic is found is 
https://github.com/NuGet/NuGet.Client/pull/2424/files#diff-0a6246a67c49154a731a0cb6a37218a0R10